### DISCUSSION TITLE
Using Minio but /api/attachments.redirect returns an AWS URL

### DISCUSSION THREAD

**Discussion by s-praveen-kumar** [outline/outline#3696] [Self Hosting]
I am hosting outline on a VPS behind Nginx proxy and using Minio for storage.
I am able to upload image to minio. But it is not displayed in outline after a refresh. Turns out the XHR request to `/api/attachments.redirect` is returning a AWS URL instead of my minio instance. Hence my images do not load. Any ideas?

Here is my configuration:
```
AWS_ACCESS_KEY_ID=xxxxxxx
AWS_SECRET_ACCESS_KEY=xxxxxxxxxxx
AWS_S3_UPLOAD_BUCKET_URL=https://sub.domain.tld
AWS_S3_UPLOAD_BUCKET_NAME=outline
AWS_S3_UPLOAD_MAX_SIZE=26214400
AWS_S3_FORCE_PATH_STYLE=true
AWS_S3_ACL=private
```

In Nginx
```	
location /outline {
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header Host $host;

		proxy_connect_timeout 300;
		proxy_http_version 1.1;
		proxy_set_header Connection "";
		chunked_transfer_encoding off;
# This is for preflight request in browser
		if ($request_method = OPTIONS) {
			add_header Access-Control-Allow-Origin '*';
			add_header Access-Control-Allow-Methods 'GET, POST, OPTIONS';
			add_header Content-Type text/plain;
			add_header Content-Length 0;
			return 204;
		}
		proxy_pass http://127.0.0.1:9000;

	}

```

**Comment by milas** [comment]
I ran into this too as well - try setting `AWS_S3_ACCELERATE_URL` to the same value as `AWS_S3_UPLOAD_BUCKET_URL`.

(It doesn't matter if you're actually using S3 Accelerate, e.g. I'm on Linode.)

---

For any of the devs reading this, I think the issue is coming from this gnarly ternary: https://github.com/outline/outline/blob/e24a5adbd50f9d99300915de77b609f999d6d833/server/utils/s3.ts#L23-L27=

It's setting `endpoint` to `null` in the case that you provide a VirtualHost style path (e.g. bucket in hostname), so when signing/generating the URL, it's falling back to the AWS SDK default value, i.e. for AWS.

Setting `AWS_S3_ACCELERATE_URL` is a workaround since that always takes precedence, but thankfully it doesn't cause any other unwanted side effects :grimacing:

**Reply by s-praveen-kumar** [reply]
Thank you @milas for your reply. I tried setting `AWS_S3_ACCELERATE_URL`. Now AWS server is no longer accessed. But I am facing another issue now.
I tried setting `AWS_S3_ACCELERATE_URL` to just my domain URL : `https://sub.domain.tld/`. Uploading doesn't work now as it doesn't call the bucket's URL. 
If I set `AWS_S3_ACCELERATE_URL` to my bucket URL : `https://sub.domain.tld/bucket`, it uploads correct. But when fetching it calls `https://sub.domain.tld/bucket/bucket` which is obviously wrong.

**Reply by Nicicalu** [reply]
@s-praveen-kumar I found something that works for me!
For me it only worked when adding the bucket name to the URL and setting `AWS_S3_FORCE_PATH_STYLE=false`:
```
AWS_REGION=eu-east-1
AWS_S3_ACCELERATE_URL=https://minio.mydomain.com/docs
AWS_S3_UPLOAD_BUCKET_URL=https://minio.mydomain.com/docs
AWS_S3_UPLOAD_BUCKET_NAME=docs
AWS_S3_FORCE_PATH_STYLE=false
```

As mentioned in [this issue](https://github.com/outline/outline/issues/2068) you shouldn't use the word `outline` in your Minio URL nor in the bucket name.

**Reply by Nicicalu** [reply]
As for my ACL on the bucket, I used the following, so that the images in the `public` folder can also be accessed (Avatar, ...)
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject",
                "s3:ListMultipartUploadParts"
            ],
            "Resource": [
                "arn:aws:s3:::my_bucket_name/public/*"
            ]
        }
    ]
}
```
